File: fioincline/lvovich/org.py
"""Склонение названий организаций — по аналогии с city.go."""

from . import types as T
from .rules import find_rule, apply_rule, ends_with
from .incline import incline_lastname
from .rules_data import (
    org_rules, org_frozen_words, org_generic_words, org_frozen_parts, org_frozen_parts_after,
)


_QUOTE_OPEN = ["\u00ab", "\u201c", '"', "'"]
_QUOTE_CLOSE = ["\u00bb", "\u201d", '"', "'"]


def split_org(name):
    """Разбивает название организации на токены, сохраняя разделители и кавычки."""
    parts = []
    buf = []
    in_quote = ""
    quote_stack = []

    def flush():
        if buf:
            parts.append("".join(buf))
            buf.clear()

    for r in name:
        ch = r
        matched = False
        # открывающая кавычка
        for qi, qo in enumerate(_QUOTE_OPEN):
            if ch == qo and in_quote == "":
                flush()
                parts.append(ch)
                in_quote = qo
                quote_stack.append(_QUOTE_CLOSE[qi])
                matched = True
                break
        if matched:
            continue
        # закрывающая кавычка
        if quote_stack and ch == quote_stack[-1]:
            flush()
            parts.append(ch)
            quote_stack.pop()
            if not quote_stack:
                in_quote = ""
            matched = True
        if matched:
            continue
        if ch == " " or ch == "-":
            flush()
            parts.append(ch)
        else:
            buf.append(ch)
    flush()
    return parts, quote_stack


def is_generic_word(str_):
    return str_.lower() in org_generic_words


def is_frozen_org(str_):
    return str_.lower() in org_frozen_words


def is_frozen_org_part(part, i, parts):
    if len(parts) > 1:
        if part.lower() in org_frozen_parts:
            return True
        for k in range(0, i):
            if parts[k].lower() in org_frozen_parts_after:
                return True
    return False


_VOWELS = "аеёиоуыэюя"
_FEMALE_ENDINGS = ("а", "я")


def gender_by_ending(word):
    if word == "":
        return T.ANDROGYNOUS
    lower = word.lower()
    for e in _FEMALE_ENDINGS:
        if lower.endswith(e):
            return T.FEMALE
    last = lower[-1]
    if last not in _VOWELS:
        return T.MALE
    return T.ANDROGYNOUS


def is_quote(s):
    if s in _QUOTE_OPEN or s in _QUOTE_CLOSE:
        return True
    if len(s) == 1 and s in ("\u00ab", "\u00bb", "\u201c", "\u201d", '"', "'"):
        return True
    return False


def decline_org(name, word_case):
    if name == "":
        return name
    if is_frozen_org(name):
        return name

    parts, _ = split_org(name)
    if len(parts) == 0:
        return name

    generic_idx = -1
    for i, part in enumerate(parts):
        part = part.strip()
        if part == "" or part == " " or part == "-":
            continue
        if is_generic_word(part):
            generic_idx = i
            break

    start_idx = 0
    if generic_idx >= 0:
        start_idx = generic_idx + 1

    if start_idx >= len(parts):
        start_idx = generic_idx
        generic_idx = -1

    out = list(parts)
    for i, part in enumerate(parts):
        part_trimmed = part.strip()
        if part_trimmed == "" or part == " " or part == "-" or is_quote(part):
            out[i] = part
            continue
        if is_frozen_org_part(part_trimmed, i, parts):
            out[i] = part
            continue
        if i == generic_idx:
            out[i] = part
            continue
        if i >= start_idx:
            if is_frozen_org(part_trimmed):
                out[i] = part
                continue
            part_gender = gender_by_ending(part_trimmed)
            rule = find_rule(part_trimmed, part_gender, org_rules, False)
            if rule is not None:
                out[i] = apply_rule(rule, part, word_case)
            else:
                if word_case == T.ACCUSATIVE and part_gender == T.MALE:
                    out[i] = part
                else:
                    res = incline_lastname(part, word_case, part_gender)
                    if res == "":
                        res = part
                    out[i] = res
        else:
            out[i] = part
    return "".join(out)


def org_in(name):
    return decline_org(name, T.PREPOSITIONAL)


def org_from(name):
    return decline_org(name, T.GENITIVE)


def org_to(name):
    return decline_org(name, T.ACCUSATIVE)


def org_decline(name, declension):
    dec = T.get_declension_const(declension)
    if dec == T.DECL_NULL:
        return name
    return decline_org(name, dec)

File: internal/lvovich/org.go
// Склонение названий организаций — по аналогии с city.go.
package lvovich

import (
	"strings"
)

// splitOrg — разбивает название организации на токены, сохраняя разделители
// (пробел, дефис, кавычки). Кавычки всех видов (« », " ") нормализуются
// во внутреннее представление, но при сборке возвращаются в исходном виде.
func splitOrg(name string) (tokens []string, quotes []string) {
	type qt struct {
		open, close string
		normal      string
	}
	quoteTypes := []qt{
		{"\u00ab", "\u00bb", "\u00ab"}, // « »
		{"\u201c", "\u201d", "\u201c"}, // " "
		{"\"", "\"", "\""},            // " (ASCII)
		{"'", "'", "'"},               // ' (ASCII)
	}

	var parts []string
	var buf []rune
	flush := func() {
		if len(buf) > 0 {
			parts = append(parts, string(buf))
			buf = buf[:0]
		}
	}

	inQuote := ""
	quoteStack := []string{}

	for _, r := range name {
		ch := string(r)
		matched := false
		// открывающая кавычка
		for _, qt := range quoteTypes {
			if ch == qt.open && inQuote == "" {
				flush()
				parts = append(parts, ch)
				inQuote = qt.normal
				quoteStack = append(quoteStack, qt.close)
				matched = true
				break
			}
		}
		if matched {
			continue
		}
		// закрывающая кавычка
		if len(quoteStack) > 0 && ch == quoteStack[len(quoteStack)-1] {
			flush()
			parts = append(parts, ch)
			quoteStack = quoteStack[:len(quoteStack)-1]
			if len(quoteStack) == 0 {
				inQuote = ""
			}
			matched = true
		}
		if matched {
			continue
		}
		if r == ' ' || r == '-' {
			flush()
			parts = append(parts, string(r))
		} else {
			buf = append(buf, r)
		}
	}
	flush()

	return parts, quoteStack
}

// isGenericWord — true, если слово — родовое (не склоняется).
func isGenericWord(str string) bool {
	lower := strings.ToLower(str)
	for _, w := range orgGenericWords {
		if w == lower {
			return true
		}
	}
	return false
}

// isFrozenOrg — true, если слово полностью несклоняемое.
func isFrozenOrg(str string) bool {
	lower := strings.ToLower(str)
	for _, w := range orgFrozenWords {
		if w == lower {
			return true
		}
	}
	return false
}

// isFrozenOrgPart — true, если часть названия не склоняется (предлог/союз).
func isFrozenOrgPart(part string, i int, parts []string) bool {
	if len(parts) > 1 {
		if isFrozen(part, orgFrozenParts) {
			return true
		}
		for k := 0; k < i; k++ {
			if isFrozen(parts[k], orgFrozenPartsAfter) {
				return true
			}
		}
	}
	return false
}

// genderByEnding — определяет грамматический род слова по окончанию
// (для выбора правила fallback при склонении значимой части названия).
func genderByEnding(word string) GenderConst {
	if word == "" {
		return Androgynous
	}
	lower := strings.ToLower(word)
	// окончания на гласную — женский/средний (не склоняются как мужской род)
	femaleEndings := []string{"а", "я"}
	for _, e := range femaleEndings {
		if strings.HasSuffix(lower, e) {
			return Female
		}
	}
	// окончания на согласную — мужской род
	vowels := "аеёиоуыэюя"
	last := lower[len(lower)-1]
	if !strings.ContainsRune(vowels, rune(last)) {
		return Male
	}
	// прочие гласные (о, е, и, у, ю, ы, э) — средний/несклоняемый
	return Androgynous
}

// declineOrg — склонение названия организации по падежу.
func declineOrg(name string, wordCase DeclenType) string {
	if name == "" {
		return name
	}

	// Проверка на полную заморозку
	if isFrozenOrg(name) {
		return name
	}

	parts, _ := splitOrg(name)
	if len(parts) == 0 {
		return name
	}

	// Найти родовое слово (первый не-разделитель, не-кавычка)
	genericIdx := -1
	for i, part := range parts {
		part = strings.TrimSpace(part)
		if part == "" || part == " " || part == "-" {
			continue
		}
		if isGenericWord(part) {
			genericIdx = i
			break
		}
	}

	// Если есть родовое слово — склоняем всё после него (значимую часть)
	// Если нет — склоняем всё название
	startIdx := 0
	if genericIdx >= 0 {
		startIdx = genericIdx + 1
	}

	// Если после родового слова ничего нет — склоняем родовое слово
	if startIdx >= len(parts) {
		startIdx = genericIdx
		genericIdx = -1
	}

	out := make([]string, len(parts))
	for i, part := range parts {
		partTrimmed := strings.TrimSpace(part)
		if partTrimmed == "" || part == " " || part == "-" || isQuote(part) {
			out[i] = part
			continue
		}
		// Разделители, предлоги, союзы — не склоняем
		if isFrozenOrgPart(partTrimmed, i, parts) {
			out[i] = part
			continue
		}
		// Родовое слово — не склоняем
		if i == genericIdx {
			out[i] = part
			continue
		}
		// Значимая часть — склоняем
		if i >= startIdx {
			// Проверка на замороженное слово
			if isFrozenOrg(partTrimmed) {
				out[i] = part
				continue
			}
			partGender := genderByEnding(partTrimmed)
			rule := FindRule(partTrimmed, partGender, orgRules, false)
			if rule != nil {
				out[i] = applyRule(*rule, part, wordCase)
			} else {
				// Для неодушевлённых мужского рода винительный = именительный
				if wordCase == Accusative && partGender == Male {
					out[i] = part
				} else {
					res := InclineLastname(part, wordCase, partGender)
					if res == "" {
						res = part
					}
					out[i] = res
				}
			}
		} else {
			out[i] = part
		}
	}
	return strings.Join(out, "")
}

// isQuote — true, если токен — кавычка.
func isQuote(s string) bool {
	switch s {
	case "\u00ab", "\u00bb", "\u201c", "\u201d", "\"", "'":
		return true
	}
	if len(s) == 1 {
		r := []rune(s)[0]
		return r == '\u00ab' || r == '\u00bb' || r == '\u201c' || r == '\u201d' || r == '"' || r == '\''
	}
	return false
}

// OrgIn — организация в предложном падеже ("в «Ромашке»").
func OrgIn(name string) string {
	return declineOrg(name, Prepositional)
}

// OrgFrom — организация в родительном падеже ("из «Ромашки»").
func OrgFrom(name string) string {
	return declineOrg(name, Genitive)
}

// OrgTo — организация в винительном падеже ("в «Ромашку»").
func OrgTo(name string) string {
	return declineOrg(name, Accusative)
}

// OrgDecline — склонение организации в указанный падеж.
// declension — строка или DeclenType.
func OrgDecline(name string, declension interface{}) string {
	dec := GetDeclensionConst(declension)
	if dec == DeclenNull {
		return name
	}
	return declineOrg(name, dec)
}